### ISSUE TITLE
How to print characters with umlaut or accents?

### ISSUE THREAD

**Issue by ldicocco** [fabienbellanger/escpos-rs#5]
Thank you for this wonderful crate. There is one thing that I don't understand. How can the Encoding be set to a specific page code of the printer? As I see that French is supported, I suppose that there is a way to do it, but I was able to do that in crates that used the original Encoding crate, I cannot finf a way to customize the used Encoding.

**Comment by travisbaars** [comment]
What sort of printer are you using? I have an old Epson and apparently it might not like UTF-8 encoding which is the default encoding here. Like you mentioned I also couldn't find a workaround in the `encoding_rs` crate. That said, you can use special characters such as the umlaut using the `custom()` option when building your `Printer::new()`. This requires knowing the Hex code for the character you wish to print. 

## Example

```rust
use escpos::printer::Printer;
use escpos::utils::*;
use escpos::{driver::*, errors::Result};

// You can define the character code via hex in a const like below.
// It's just an array holding the character codes you wish to send to the printer.
const UMLAUT_LOWER_A: &[u8] = &[0x84];

fn main() -> Result<()>{
    let driver = NetworkDriver::open(PRINTER_IP, 9100)?;
    Printer::new(driver, Protocol::default())
        .debug_mode(Some(DebugMode::Dec))
        .init()?
        .custom(UMLAUT_LOWER_A)?
        // You could also just use the codes directly in the .custom() command. 
        // (even multiple codes).
        .custom(&[0x84, 0x84])?
        .feeds(4)?
        .print_cut()?;

    Ok(())
}
```

This example prints out `ä`.

I have confirmed this to work with my Epson printer. I assume this works on others as well.

I used [this](https://reference.epson-biz.com/modules/ref_charcode_en/index.php?content_id=10) chart to help me find my desired character code. The chart is read column/row.

Hope this helps!

**Comment by ldicocco** [comment]
Thank you @travisbaars  for your answer. In effect I am already using the workaround, and I have a small utility function for this:
`use std::collections::HashMap;

use map_macro::hash_map;
use unicode_segmentation::UnicodeSegmentation;

pub fn encode(input: &str) -> Vec<u8> {
  let mut escpos_text: Vec<u8> = Vec::new();
  input
    .graphemes(true)
    .for_each(|grapheme| escpos_text.extend_from_slice(&grapheme_to_escpos_bytes(grapheme)));

  escpos_text
}

fn map_char(char: &str) -> Option<u8> {
  let map: HashMap<&str, u8> = hash_map! {
    "Ä" => 142,
    "ä" => 132,
    "Ö" => 153,
    "ö" => 148,
    "Ü" => 154,
    "ü" => 129,
    "ß" => 225,
    "Ø" => 157,
    "ø" => 155,
    "°" => 248,
    "²" => 253,
  };

  map.get(char).copied()
}

fn grapheme_to_escpos_bytes(grapheme: &str) -> Vec<u8> {
  match map_char(grapheme) {
    Some(value) => vec![value],
    _ => grapheme.as_bytes().to_owned(),
  }
}
`
so that I can use something like:
`.custom(encode(&name).as_bytes())?`

It still looks a bit strange to me that in this crate, that in my opinion is by far the best one to support esc/pos, `.write` in fact support only basic ASCII, as the 127 basic ASCII char are the only ones where the UTF-8 encoding is the same.
At least in Germany and Italy where I work, I have never seen in any shop a thermal printer that supports UTF-8 so, seeing that the library creator is French, I expected a direct support for international out of the box, and thought that it was just me that couldn't see it.

**Comment by travisbaars** [comment]
I agree with you on the fact that this is the best crate for ESC/POS. This was the only one that I was able to find that had network capabilities which was very important, and it worked right out of the box. The lack of ASCII support is odd, but I suspect it's due to the use of the `encoding_rs` crate. It seems to not support ASCII, apparently. 

 Now I was messing around with using encoding options and found something interesting. The only encoding that worked was the default UTF-8. None of the other encodings made it past the match in  the `Encoder::encode()` method:

```rust
/// Encode string into the right codec
    pub(crate) fn encode(&self, data: &str) -> Result<Vec<u8>> {
        match self.codec.can_encode_everything() {
            true => Ok(self.codec.encode(data).0.into()),
            false => Err(crate::errors::PrinterError::Input(format!(
                "invalid {}",
                self.codec.name()
            ))),
        }
    }
```

This match, `match self.codec.can_encode_everything() {...}` ,only passes if the output encoding is UTF-8.:
```
Checks whether the output encoding of this encoding can encode every char. (Only true if the output encoding is UTF-8.)
```

So I think in order to get proper support for the alternate encodings there would need to be big change to what is currently there. 

In my opinion, this definitely seems like something that should be changed. 
Now I don't speak for the creator, but I think this is something that could be fixed with some work. We could maybe use what you came up with as the starting point for the solution to this problem.

**Comment by travisbaars** [comment]
I can also help see if I can find a good solution.

**Comment by fabienbellanger** [comment]
I've added a branch `feature/codePage` that provides a quick solution.
It's certainly not the most efficient and elegant but it seems to work. I've only implemented the PC437 and PC858 page code.
I've added an example in `examples/page_codes.rs`.

**Comment by travisbaars** [comment]
I ran your example and it seemed to work for the PC858 PageCode but it messed up some of the PC437 PageCode.

Attached an image of what my printer spit out:
 
![image](https://github.com/fabienbellanger/escpos-rs/assets/73616094/a3c4a343-8434-4dea-8d1d-54ead8b6d7b1)

**Comment by fabienbellanger** [comment]
Yes I have the same result.

The characters €, À and Ô are not available in the PC437 page code, which is why the printer cannot print them correctly (the aim of the example was to show that the choice of page code was important for the correct display of the characters).

**Comment by travisbaars** [comment]
> Yes I have the same result.
> 
> The characters €, À and Ô are not available in the PC437 page code...

I think PC437 was US so that would make sense.

**Comment by fabienbellanger** [comment]
If it's OK for you, I will publish a new release (`0.7.0`).

**Comment by fabienbellanger** [comment]
I've just published version `0.7.0`.

Thanks for your help 😊
